Sources/MapboxDirections/DirectionsOptions.swift

```diff
@@ -122,10 +123,133 @@ open class DirectionsOptions: Codable {
      */
-    required public init(waypoints: [Waypoint], profileIdentifier: ProfileIdentifier? = nil) {
+    required public init(waypoints: [Waypoint], profileIdentifier: ProfileIdentifier? = nil, queryItems: [URLQueryItem]? = nil) {
         self.waypoints = waypoints
         self.profileIdentifier = profileIdentifier ?? .automobile
+        
+        guard let queryItems = queryItems else {
+            return
+        }
+        
+        let mappedQueryItems = Dictionary<String, String>(uniqueKeysWithValues: queryItems.compactMap {
+            guard let value = $0.value else { return nil }
+            return ($0.name, value)
+        })
+        
+        if let mappedValue = mappedQueryItems[CodingKeys.shapeFormat.stringValue],
+           let shapeFormat = RouteShapeFormat(rawValue: mappedValue) {
+            self.shapeFormat = shapeFormat
+        }
+        if let mappedValue = mappedQueryItems[CodingKeys.routeShapeResolution.stringValue],
+           let routeShapeResolution = RouteShapeResolution(rawValue: mappedValue) {
+            self.routeShapeResolution = routeShapeResolution
+        }
+        if mappedQueryItems[CodingKeys.includesSteps.stringValue] == "true" {
+            self.includesSteps = true
+        }
+        if let mappedValue = mappedQueryItems[CodingKeys.locale.stringValue] {
+            self.locale = Locale(identifier: mappedValue)
+        }
+        if mappedQueryItems[CodingKeys.includesSpokenInstructions.stringValue] == "true" {
+            self.includesSpokenInstructions = true
+        }
+        if let mappedValue = mappedQueryItems[CodingKeys.distanceMeasurementSystem.stringValue],
+           let measurementSystem = MeasurementSystem(rawValue: mappedValue) {
+            self.distanceMeasurementSystem = measurementSystem
+        }
+        if mappedQueryItems[CodingKeys.includesVisualInstructions.stringValue] == "true" {
+            self.includesVisualInstructions = true
+        }
+        if let mappedValue = mappedQueryItems[CodingKeys.attributeOptions.stringValue],
+           let attributeOptions = AttributeOptions(descriptions: mappedValue.components(separatedBy: ",")) {
+            self.attributeOptions = attributeOptions
+        }
+        if let mappedValue = mappedQueryItems["waypoints"] {
+            let indicies = mappedValue.components(separatedBy: ";").compactMap { Int($0) }
+            if !indicies.isEmpty {
+                waypoints.enumerated().forEach {
+                    $0.element.separatesLegs = indicies.contains($0.offset)
+                }
+            }
+        }
+        
+        let waypointsData = [mappedQueryItems["approaches"]?.components(separatedBy: ";"),
+                             mappedQueryItems["bearings"]?.components(separatedBy: ";"),
+                             mappedQueryItems["radiuses"]?.components(separatedBy: ";"),
+                             mappedQueryItems["waypoint_names"]?.components(separatedBy: ";"),
```

Comment:
At a glance, I don’t see any glaring logic errors in this conversion. However, I’m a bit concerned about the maintainability of this code as the Directions API evolves. Let’s at least increase the code coverage level on these additions by expanding the unit tests.

As possible tail work, it would be nice to find a way to fold it into the existing Codable implementation, which expects a JSON representation of the same data that’s pretty consistent with the URL query parameters following #593. Maybe we could improve maintainability by first converting the `URLQueryItem`s into the JSON representation very literally, then decoding that JSON data. If there are any roundtripping issues that way, they would also affect codability.

Reply:
Covered all of these by unit test check